### PR TITLE
Add parser and combiner to parse files of /sys/devices/system/cpu/vulnerabilities/*

### PR DESCRIPTION
**Pull request by xhe123**
Add Parser & Combiner for file ``/sys/devices/system/cpu/vulnerabilities/*`` outputs

* This module provides:
  - parser of CpuVulns file ``/sys/devices/system/cpu/vulnerabilities/*`` 
  - combiner of CpuVulnsBranch for multiple parsers of CpuVulns

Signed-off-by: Xiaonan He <xhe@redhat.com>

### CODE REVIEW

**Comment by xhe123** [comment]
@matysek @skontar may I have your review on this parser? Thanks!

**Comment by skontar** [comment]
I think it would be better to use existing parser from our repository. You can move it to insights-core. It just needs documentation as it was not expected to be used outside of Product Security rules.

https://gitlab.cee.redhat.com/insights-rules/insights-prodsec-rules/blob/master/prodsec/parsers/cpu_vulns.py

**Comment by xhe123** [comment]
@skontar Thanks for your review and suggestion.

I added the classes that prodsec-rules required, here is the test result:
```
(.python3) [xiaonan@dhcp-136-18 insights-core]$ pytest -svk cpu_vulns --cov=insights/parsers |grep cpu_vulns
/home/xiaonan/wk/code/gitlab/insights-core/insights/combiners/nfs_exports.py:73: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  if isinstance(nfsexportsd, collections.Iterable):
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_3 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_smt PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_smt_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_smt_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_3 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_4 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel7 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel6 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_exp2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_documentation PASSED
insights/parsers/cpu_vulns.py                           34      0     18      0   100%
```

**Comment by xhe123** [comment]
code coverage is 100% now.

```
(.python3) [xiaonan@dhcp-136-18 insights-core]$ pytest -svk cpu_vulns --cov=insights/parsers |grep cpu_vulns
/home/xiaonan/wk/code/gitlab/insights-core/insights/combiners/nfs_exports.py:73: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  if isinstance(nfsexportsd, collections.Iterable):
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_3 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_smt PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_smt_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_3 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_4 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_mds_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel7 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel6 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_documentation PASSED
insights/parsers/cpu_vulns.py                           29      0     16      0   100%

```

**Comment by skontar** [comment]
I did not do a complete review, just a quick look. I may be mistaken, but the implementation does not seem to be correct. I do not think you will get correct values on the system which contains all the vulnerability files, not just one.

The spec takes all the files in the `/sys/devices/system/cpu/vulnerabilities/` directory. The filename checks need to be part of the classes as seen in the code located in ProdSec repository.

Do you have any reason why you did not use the already tested code and tests as they are in the ProdSec repository?

**Comment by xhe123** [comment]
@skontar YES, your concern is reasonable. I found we have existing four independent SPECs files in insights-core/insights/specs/default.py, your comment bring me a good idea, if we have the dependent specs already, it will be better to use these separate SPECs, then I don't have to check file_name in code any more.  

However actually, there are six specs are required if I want to code the parser with dependent specs. So maybe I can implement this parser with two steps: 
1) complete current parser for the four dependent SPECs firstly
```
    cpu_vulns_meltdown = simple_file("sys/devices/system/cpu/vulnerabilities/meltdown")
    cpu_vulns_spectre_v1 = simple_file("sys/devices/system/cpu/vulnerabilities/spectre_v1")
    cpu_vulns_spectre_v2 = simple_file("sys/devices/system/cpu/vulnerabilities/spectre_v2")
    cpu_vulns_spec_store_bypass = simple_file("sys/devices/system/cpu/vulnerabilities/spec_store_bypass") 
```
2) we open a SPEC JIRA to add new specs for **l1tf** and **mds**, add another two parser for **l1tf** and **mds** in another PR. (I may open the JIRA if you are busy recently)
```
    cpu_vulns_l1tf = simple_file("sys/devices/system/cpu/vulnerabilities/l1tf")
    cpu_vulns_mds = simple_file("sys/devices/system/cpu/vulnerabilities/mds")
```
Do you think it could be a good solution to me here? Thanks!

**Comment by skontar** [comment]
I now see what is causing the confusion, sorry about that. Specs for specific files should have been removed some time ago. The only spec which should exist now is:
```
cpu_vulns = glob_file("sys/devices/system/cpu/vulnerabilities/*")
```
This is a `glob_file` spec which is different to `simple_file` and needs to be handled differently. It also will gather any vulnerability files added in the future.

I think that the best course of action would be for my team to clean up and port the parsers to the core as we have much more code which depends on it and which can potentially break. We could have it done early next week.

What do you think?

**Comment by xhe123** [comment]
@skontar honestly, I would like to keep the existing `simple_spec`, it makes the parser simple and clear, the parser parser(s) do not need to be intertwined and impact in any cases, we don't really have to deal with any particular case for the mixed specs.  

I don't know the purpose of using `glob_file`, let us raise this question to insights framework sync meeting later to get some advise?

**Comment by skontar** [comment]
@xhe123 I sincerely disagree. `glob_file` is the correct way how to gather information from vulnerability files. `simple_file` specs for the same files are past relic which should have been removed some time ago (I think they are currently unused as they should). Parsers written in ProdSec repository are simple and handle the `glob_file` spec correctly. I do not think that duplicating specs and parsers is the correct way forward.

**Comment by kgrant-rh** [comment]
@xhe123 I think @skontar 's concern, and mine too, is that a system will have several of these files:

> [kgrant@localhost ~]$ ls /sys/devices/system/cpu/vulnerabilities/
> l1tf  mds  meltdown  spec_store_bypass  spectre_v1  spectre_v2

Since the parsers will process each file from the glob list, one at a time, it means that in this case every parser will have the contents of the last file, in this case spectre_v2. That is the purpose for the somewhat convoluted `SkipException` in our parsers for those files -- they make sure they only take content from the right file.

**Comment by csams** [comment]
@xhe123 I agree that the ideal technical solution is to have a spec and parser for each file since they represent different things and can have their own formats. However, these files represent particularly critical issues all related to the same area, and in this case I agree with the team's decision to trade the ideal technical solution for one that removes "new spec" workflow that would be a formality anyway.

@skontar @kgrant-rh I recommend creating a new public combiner and migrating the existing prodsec parser with whatever changes you feel necessary into its module as a private implementation detail.

Rules should depend only on the new combiner, and we'll be in a better position to change everything else later if we need to.

**Comment by skontar** [comment]
@xhe123 @csams Thanks for your insights and patience. We are tracking the issue under https://bugzilla.redhat.com/show_bug.cgi?id=1753695 . Thanks!

**Comment by xhe123** [comment]
@csams @kgrant-rh @skontar  Thanks for your help and suggestions on this PR. I will roll back code to the previous version that is based on glob_file. Then invite Stanishlav and Martin to help review it later.

File: insights/combiners/cpu_vulns_branch.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@combiner(CpuVulns)
+class CpuVulnsBranch(object):
+    """
+    This combiner provides an interface to CPU vulnerabilities parsers.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+
+    @property
```
[the remark this answers is not in the document]
Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: yes, I removed `property` .

File: insights/combiners/cpu_vulns_branch.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import os
+
+
+@combiner(CpuVulns)
+class CpuVulnsBranch(object):
+    """
+    This combiner provides an interface to CPU vulnerabilities parsers.
+    """
+
+    def __init__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: yes, I updated code as you suggested.

File: insights/combiners/cpu_vulns_branch.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+CpuVulnsBranch - combiner for cpu vulnerabilities:
+==================================================
+
+cpu vulnerabilities includes:
+    * CpuVulnsBranch - file /sys/devices/system/cpu/vulnerabilities/*
+
+Examples:
+    >>> type(cvb)
+    <class 'insights.combiners.cpu_vulns_branch.CpuVulnsBranch'>
+    >>> cvb.get_data
```
[the remark this answers is not in the document]
Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: yes, I changed the test as you suggested

File: insights/combiners/cpu_vulns_branch.py [hunk 4]
```diff
@@ -8,35 +8,24 @@
...
-        for arg in self.args:
-            fdict = {}
-            if arg.path and arg.content[0]:
-                fname = os.path.basename(arg.path)
-                fdict[fname] = arg.content[0]
-                flist.append(fdict)
-        return flist
+    def __init__(self, cpu_vulns):
+        for cv in cpu_vulns:
+            self.update({cv.path: cv.value}) if cv.path else None
```
Comment (xiangce) [inline insights/combiners/cpu_vulns_branch.py]: You should use `cv.file_name` instead of `dv.path`

File: insights/parsers/cpu_vulns.py [hunk 5]
```diff
@@ -26,6 +26,11 @@ class CpuVulns(Parser):
         SkipException: When file content is empty
 
     """
+
+    def __init__(self, *args, **kwards):
```
Comment (xiangce) [inline insights/parsers/cpu_vulns.py]: It's not necessary to add this `__init__.py`

**Comment by xhe123** [comment]
parser and combiner are 100% code coverage

Here are the test results of parser
```
(.python3) [xiaonan@dhcp-136-18 insights-core]$ pytest -svk cpu_vulns --cov=insights/parsers |grep cpu_vulns
/home/xiaonan/wk/code/gitlab/insights-core/insights/combiners/nfs_exports.py:73: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  if isinstance(nfsexportsd, collections.Iterable):
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_meltdown PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spectre_v1 PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spectre_v2 PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spec_store_bypass PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_l1tf PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_mds PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_integration PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_exp PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_x86_enabled_documentation PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_meltdown_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_2 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_3 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spec_store_bypass_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v1_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel7 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_rhel6 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_spectre_v2_exp1 PASSED
insights/parsers/tests/test_cpu_vulns.py::test_cpu_vulns_documentation PASSED
insights/parsers/cpu_vulns.py                           14      0      4      0   100%
```
Here are the test results of combiner
```
(.python3) [xiaonan@dhcp-136-18 insights-core]$ pytest -svk cpu_vulns_br --cov=insights/combiners |grep cpu_vulns_branch
/home/xiaonan/wk/code/gitlab/insights-core/insights/combiners/nfs_exports.py:73: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  if isinstance(nfsexportsd, collections.Iterable):
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_meltdown PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spectre_v1 PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spectre_v2 PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_spec_store_bypass PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_l1tf PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_comb_mds PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_integration PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_values_exp PASSED
insights/combiners/tests/test_cpu_vulns_branch.py::test_x86_enabled_documentation PASSED
insights/combiners/cpu_vulns_branch.py                10      0      6      0   100%
 
```

File: insights/combiners/cpu_vulns_branch.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+"""
+CpuVulnsBranch - combiner for cpu vulnerabilities:
```
Comment (skontar) [inline insights/combiners/cpu_vulns_branch.py]: Why is the name `CpuVulnsBranch`?

File: insights/parsers/tests/test_cpu_vulns.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+# The kernel is unable to reliably determine whether Hyper-Threading is enabled when running in a virtual environment. Updated host kernel packages, updated host qemu packages with proper configuration to pass through the host CPU type to the guest, and updated guest kernel packages will show the following status inside of the virtual environment:
+INPUT_MDS_5 = """
+Mitigation: Clear CPU buffers; SMT Host state unknown
+""".strip()
+
+
+# CpuVulnsMeltdown
+def test_cpu_vulns_meltdown():
+    spectre = CpuVulns(context_wrap(INPUT_MELTDOWN, path='meltdown'))
```
Comment (skontar) [inline insights/parsers/tests/test_cpu_vulns.py]: Should be real path we expect: `path='/sys/devices/system/cpu/vulnerabilities/meltdown'`

This applies for all tests.

Comment (xhe123) [reply insights/parsers/tests/test_cpu_vulns.py]: Yes, it sounds make sense. Let me modify it to real paths.

File: insights/combiners/cpu_vulns_branch.py (hunk 6 above)
Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: no much good name to me, may I have your advice on this name?

Comment (skontar) [reply insights/combiners/cpu_vulns_branch.py]: Maybe the same name as for the parser? `CpuVulns` would not make a naming conflict as they are both in different modules. Or maybe `CpuVulnsCombined` or `CpuVulnsAll` as there is `HttpdConfAll` which has similar function.

Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: CpuVulnsCombined sound good to me, let me modify it. Thanks :)

File: insights/parsers/cpu_vulns.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+
+@parser(Specs.cpu_vulns)
+class CpuVulns(Parser):
+    """
+    Base class for current parser
+
+    Attributes:
+        value (str): the result parsed
+        path (str): the name of file parsed
```
Comment (xiangce) [inline insights/parsers/cpu_vulns.py]: Need to remove this `path`

Comment (xhe123) [reply insights/parsers/cpu_vulns.py]: yes, let me remove it.

File: insights/combiners/cpu_vulns_branch.py (hunk 6 above)
Comment (xhe123) [reply insights/combiners/cpu_vulns_branch.py]: @skontar finally, I updated the combiner name with CpuVulnsAll. will you please review this PR again? Thanks!

File: insights/parsers/cpu_vulns.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+"""
+Parsers for cpu vulnerabilities file output
+===========================================
+
+This module provides the following parser:
+
+CpuVulns - ``/sys/devices/system/cpu/vulnerabilities/*``
+"""
+
+from __future__ import division
```
Comment (matysek) [inline insights/parsers/cpu_vulns.py]: Just curious, is there any reason for this import statement? If so please add comment about it.

Comment (xhe123) [reply insights/parsers/cpu_vulns.py]: It was for python 2 on some digital computation, it doesn't impact much in current parser. I can remove it. Thanks for your comment.

**Comment by xhe123** [comment]
@skontar Does the latest code look ok to you? waiting for your response here. Thanks!

**Comment by skontar** [comment]
Looks good. @matysek tested it with our code, so it works well. Thanks for all that work!

**Pull request merged**